src/bin/drag_and_drop_listbox_model.rs

```diff
@@ -0,0 +1,653 @@
+    // Add custom css style
+    let css_provider = gtk::CssProvider::new();
+    css_provider.load_from_data(CSS.as_bytes()).unwrap();
+    gtk::StyleContext::add_provider_for_screen(
+        gdk::Screen::get_default().as_ref().unwrap(),
+        &css_provider,
+        gtk::STYLE_PROVIDER_PRIORITY_APPLICATION,
+    );
+
+    let window = gtk::ApplicationWindow::new(application);
+
+    window.set_title("ListBox Model Sample");
+    window.set_border_width(10);
+    window.set_position(gtk::WindowPosition::Center);
+    window.set_default_size(320, 480);
+
+    let window_weak = window.downgrade();
+
+    let vbox = gtk::Box::new(gtk::Orientation::Vertical, 5);
+
+    // Create our list store and specify that the type stored in the
+    // list should be the RowData GObject we define at the bottom
+    let model = gio::ListStore::new(RowData::static_type());
+
+    // And then create the UI part, the listbox and bind the list store
+    // model to it. Whenever the UI needs to show a new row, e.g. because
+    // it was notified that the model changed, it will call the callback
+    // with the corresponding item from the model and will ask for a new
+    // gtk::ListBoxRow that should be displayed.
+    //
+    // The gtk::ListBoxRow can contain any possible widgets.
+    let listbox = gtk::ListBox::new();
+    listbox.set_valign(gtk::Align::Start);
+    // Create target for drag and drop if the listbox is a drag_area
+    let targets = vec![gtk::TargetEntry::new(
+        "CHANGE_POSITION",
+        // This seams to have no effect
+        gtk::TargetFlags::SAME_APP,
+        0,
+    )];
+    // Create a communication channel between the dnd source and destination
+    let channel = DndChannel::new(Cell::new(None));
+    let channel_weak = Rc::downgrade(&channel);
+
+    let model_weak = model.downgrade();
+    listbox.drag_dest_set(gtk::DestDefaults::ALL, &targets, gdk::DragAction::MOVE);
+    listbox.connect_drag_data_received(clone!(model_weak =>
+                                           move |widget, _context, _x, y, _s, _info, _time| {
+                                               let model = upgrade_weak!(model_weak);
+                                               if let Some(data) = channel.take() {
+                                                   if let Some(source) = model.get_object(data.position) {
+                                                       if let Some(ref list) = widget.downcast_ref::<gtk::ListBox>() {
+                                                           if let Some(row) = list.get_row_at_y(y) {
+                                                           model.remove(data.position);
+                                                           let index = row.get_index() as u32;
+                                                           let alloc = row.get_allocation();
+
+                                                           let index = if y < alloc.y + alloc.height/2 {
```

Comment:
I believe you want to run `cargo fmt` once :)

Reply:
Still seems like a good idea :)

Reply:
apparently rustfmt doesn't fix the spaces, actually it doesn't change the file add all

Reply:
That's curious. Don't worry about that then, I'll take a look once the code is done